cpp/src/Slice/Parser.cpp

```diff
@@ -1067,9 +1061,9 @@ Slice::Contained::getDeprecationReason() const
 }
 
 Slice::Contained::Contained(const ContainerPtr& container, string name)
-    : SyntaxTreeBase(container->unit()),
-      _container(container),
-      _name(std::move(name))
+    : _container(container),
+      _name(std::move(name)),
+      _unit(container->unit())
```

Comment:
I am puzzled by this code: instead of keeping _unit, you could just return _container->unit(), and then you don't need any destroy, since destroy() doesn't clear _container.

Reply:
I hadn't even thought of that.
That's a good idea.

Reply:
There is a downside to this: we would need to 'artificially' remove `const` from some functions.

`unit()` is a non-const function because we need the overload of `shared_from_this` which returns a non-const ptr.
Whereas accessing the `_unit` field _is_ const. And we do have const functions which use this field instead of calling `unit()`, namely `mappedScope`, `mappedName`, etc. Which in turn are used by other const functions.

I'd rather just leave this field be.
When we switch to using `weak_ptr`, all these destroy will go away anyways.

Reply:
The unit() accessor should be const.

> unit() is non-const because we need the overload of shared_from_this 

You can cast-away constness in this case, as in:
```cpp
const_cast<Unit*>(this)->shared_from_this();
```

Reply:
I knew I had this ability, but wasn't sure if it was 'safe' to do.
Fixed!